File: client/java/client/src/com/bj58/spat/gaea/client/communication/socket/CSocket.java
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package com.bj58.spat.gaea.client.communication.socket;

import java.net.Socket;
import java.nio.ByteBuffer;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.channels.SocketChannel;
import java.util.concurrent.ConcurrentHashMap;
import java.nio.channels.NotYetConnectedException;

import com.bj58.spat.gaea.client.configuration.commmunication.SocketPoolProfile;
import com.bj58.spat.gaea.client.utility.AutoResetEvent;
import com.bj58.spat.gaea.client.utility.logger.ILog;
import com.bj58.spat.gaea.client.utility.logger.LogFactory;
import com.bj58.spat.gaea.protocol.exception.DataOverFlowException;
import com.bj58.spat.gaea.protocol.exception.ProtocolException;
import com.bj58.spat.gaea.protocol.exception.TimeoutException;
import com.bj58.spat.gaea.protocol.sfp.v1.SFPStruct;
import com.bj58.spat.gaea.protocol.utility.ByteConverter;
import com.bj58.spat.gaea.protocol.utility.ProtocolConst;

/**
 * CSocket
 *
 * @author Service Platform Architecture Team (spat@58.com)
 */
public class CSocket {

    private static final ILog logger = LogFactory.getLogger(CSocket.class);
   
    private byte[] DESKey;//DES密钥
    private boolean rights;//是否启用认证
    private Socket socket;
    private ScoketPool pool;
    private SocketChannel channel;
    private ByteBuffer receiveBuffer, sendBuffer;
    private SocketPoolProfile socketConfig;
    private boolean _inPool = false;
    private boolean _connecting = false;
    private DataReceiver dataReceiver = null;
    private boolean waitDestroy = false;
    private final Object sendLockHelper = new Object();
    private final Object receiveLockHelper = new Object();
    private CByteArrayOutputStream receiveData = new CByteArrayOutputStream();
    private ConcurrentHashMap<Integer, WindowData> WaitWindows = new ConcurrentHashMap<Integer, WindowData>();

    protected CSocket(InetSocketAddress endpoint, ScoketPool _pool, SocketPoolProfile config) throws IOException {
        this.socketConfig = config;
        this.pool = _pool;
        channel = SocketChannel.open(endpoint);
        channel.configureBlocking(false);
        socket = channel.socket();
        channel.socket().setSendBufferSize(config.getSendBufferSize());
        channel.socket().setReceiveBufferSize(config.getRecvBufferSize());
        receiveBuffer = ByteBuffer.allocate(config.getBufferSize());
        sendBuffer = ByteBuffer.allocate(config.getMaxPakageSize());
        _connecting = true;
        dataReceiver = DataReceiver.instance();
        dataReceiver.RegSocketChannel(this);
        
        logger.info("MaxPakageSize:" + config.getMaxPakageSize());
		logger.info("SendBufferSize:" + config.getSendBufferSize());
		logger.info("RecvBufferSize:" + config.getRecvBufferSize());
        logger.info("create a new connection :" + this.toString());
    }
    
    public int send(byte[] data) throws IOException, Throwable {
        try {
            synchronized (sendLockHelper) {
                int pakageSize = data.length + ProtocolConst.P_END_TAG.length;
                if (sendBuffer.capacity() < pakageSize) {
                    throw new DataOverFlowException("数据包(size:" + pakageSize + ")超过最大限制,请修改或增加配置文件中的<SocketPool maxPakageSize=\"" + socketConfig.getMaxPakageSize() + "\"/>节点属性！");
                }
          
	            int count = 0;
	    		sendBuffer.clear();
	    		sendBuffer.put(data);
	    		sendBuffer.put(ProtocolConst.P_END_TAG);
	    		sendBuffer.flip();
	    		
	    		int retryCount = 0;
	    		while(sendBuffer.hasRemaining()) {
	    			count += channel.write(sendBuffer);
	    			
	    			if(retryCount++ > 10) {
	    				throw new Exception("retry write count(" + retryCount + ") above 10");
	    			}
	    		}
	    		return count;
            }
        } catch (IOException ex) {
            _connecting = false;
            throw ex;
        } catch (NotYetConnectedException ex) {
            _connecting = false;
            throw ex;
        }
    }

    public byte[] receive(int sessionId, int queueLen) throws IOException, TimeoutException, Exception {
        WindowData wd = WaitWindows.get(sessionId);
        if (wd == null) {
            throw new Exception("Need invoke 'registerRec' method before invoke 'receive' method!");
        }
        AutoResetEvent event = wd.getEvent();
        int timeout = getReadTimeout(socketConfig.getReceiveTimeout(), queueLen);
        if (!event.waitOne(timeout)) {
            throw new TimeoutException("Receive data timeout or error!timeout:" + timeout + "ms,queue length:" + queueLen);
        }
        byte[] data = wd.getData();
        int offset = SFPStruct.Version;
        int len = ByteConverter.bytesToIntLittleEndian(data, offset);
        if (len != data.length) {
            throw new ProtocolException("The data length inconsistent!datalen:" + data.length + ",check len:" + len);
        }
        return data;
    }

    public void registerRec(int sessionId) {
        AutoResetEvent event = new AutoResetEvent();
        WindowData wd = new WindowData(event);
        WaitWindows.put(sessionId, wd);
    }

    public void unregisterRec(int sessionId) {
        WaitWindows.remove(sessionId);
    }

    public void close() {
        pool.release(this);
    }

    public void dispose() throws Exception {
        dispose(false);
    }

    public void dispose(boolean flag) {
        if (flag) {
            logger.warning("destory a connection");
            try {
                dataReceiver.UnRegSocketChannel(this);
            } finally {
                pool.destroy(this);
            }
        } else {
            close();
        }
    }

    protected void disconnect() throws IOException {
        if (channel != null) {
            channel.close();
        }
        _connecting = false;
    }

    private int getReadTimeout(int baseReadTimeout, int queueLen) {
        if (!socketConfig.isProtected()) {
            return baseReadTimeout;
        }
        if (queueLen <= 0) {
            queueLen = 1;
        }
        int result = baseReadTimeout;
        int flag = (queueLen - 100) / 10;
        if (flag >= 0) {
            if (flag == 0) {
                flag = 1;
            }
            result = baseReadTimeout / (2 * flag);
        } else if (flag < -7) {
            result = baseReadTimeout - ((flag) * (baseReadTimeout / 10));
        }

        if (result > 2 * baseReadTimeout) {
            result = baseReadTimeout;
        } else if (result < 5) {
            result = 5; //min timeout is 5ms
        }
        if (queueLen > 50) {
            logger.warn("-----IsProtected:true,queueLen:" + queueLen + ",timeout:" + result + ",baseReadTimeout:" + baseReadTimeout);
        }
        return result;
    }
    
    private boolean handling = false;

    protected void frameHandle() throws IOException, InterruptedException {
        if (handling) {
            return;
        }
        synchronized (receiveLockHelper) {
            handling = true;
            try {
                if (waitDestroy && isIdle()) {
                    logger.info("Shrinking the connection:" + this.toString());
                    dispose(true);
                    return;
                }
                int index = 0;
                receiveBuffer.clear();
                try {
                    channel.read(receiveBuffer);
                } catch (IOException ex) {
                    _connecting = false;
                    throw ex;
                } catch (NotYetConnectedException e) {
                    _connecting = false;
                    throw e;
                }
                receiveBuffer.flip();
                if (receiveBuffer.remaining() == 0) {
                    return;
                }
                
                while (receiveBuffer.remaining() > 0) {
                    byte b = receiveBuffer.get();
                    receiveData.write(b);
                    if (b == ProtocolConst.P_END_TAG[index]) {
                        index++;
                        if (index == ProtocolConst.P_END_TAG.length) {
                            byte[] pak = receiveData.toByteArray(0, receiveData.size() - ProtocolConst.P_END_TAG.length);
                            int pSessionId = ByteConverter.bytesToIntLittleEndian(pak, SFPStruct.Version + SFPStruct.TotalLen);
                            WindowData wd = WaitWindows.get(pSessionId);
                            if (wd != null) {
                                wd.setData(pak);
                                wd.getEvent().set();
                            }
                            index = 0;
                            receiveData.reset();
                            continue;
                        }
                    } else if (index != 0) {
                    	if(b == ProtocolConst.P_END_TAG[0]) {
                    		index = 1;
                    	} else {
                    		index = 0;
                    	}
                    }
                }
            } finally {
                handling = false;
            }
        }
    }

    @Override
    protected void finalize() throws Throwable {
        try {
            if (_connecting || (channel != null && channel.isOpen())) {
                dispose(true);
            }
        } catch (Throwable t) {
            logger.error("Pool Release Error!:", t);
        } finally {
            super.finalize();
        }
    }

    public boolean connecting() {
        return _connecting;
    }

    protected boolean inPool() {
        return _inPool;
    }

    protected void setInPool(boolean inPool) {
        _inPool = inPool;
    }

    protected SocketChannel getChannle() {
        return channel;
    }

    /*
     * 该链接是否是空闲状态
     */
    protected boolean isIdle() {
        return !(WaitWindows.size() > 0);
    }

    protected void waitDestroy() {
        this.waitDestroy = true;
    }
    
    public boolean isRights() {
		return rights;
	}

	public void setRights(boolean rights) {
		this.rights = rights;
	}
	
	public byte[] getDESKey() {
		return DESKey;
	}

	public void setDESKey(byte[] dESKey) {
		DESKey = dESKey;
	}
    
    @Override
    public String toString() {
        try {
            return (socket == null) ? "" : socket.toString();
        } catch (Throwable ex) {
            return "Socket[error:" + ex.getMessage() + "]";
        }
    }
}

File: client/net/client/Com.Bj58.Spat.Gaea.Client/Communication/Socket/CSocket.cs
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
//----------------------------------------------------------------------
//<Copyright company="58.com">
//      Team:SPAT
//      Blog:http://blog.58.com/spat/  
//      Website:http://www.58.com
//</Copyright>
//-----------------------------------------------------------------------
using System;
using System.IO;
using System.Net;
using System.Linq;
using System.Threading;
using System.Collections;
using System.Net.Sockets;
using System.Diagnostics;
using System.Collections.Generic;
using Com.Bj58.Spat.Gaea.Client.Exceptions;
using Com.Bj58.Spat.Gaea.Client.Utility.Helper;
using Com.Bj58.Spat.Gaea.Client.Utility.Logger;
using Com.Bj58.Spat.Gaea.Client.Communication.Protocol.SFP;
using Com.Bj58.Spat.Gaea.Client.Configuration.Commmunication;

namespace Com.Bj58.Spat.Gaea.Client.Communication.Socket
{
    internal class CSocket : IDisposable
    {
        private LoggerBase logger = LoggerBase.GetLogger();
        private System.Net.Sockets.Socket socket;
        private SocketPool pool;
        private IPEndPoint endpoint;
        private BufferedStream inputStream;
        private SocketPoolProfile socketConfig;
        private Dictionary<int, WindowData> ReceiveWindows = new Dictionary<int, WindowData>();
        public CSocket(IPEndPoint _endPoint, SocketPool _pool, SocketPoolProfile config)
        {
            socketConfig = config;
            endpoint = _endPoint;
            pool = _pool;
            System.Net.Sockets.Socket _socket = new System.Net.Sockets.Socket(_endPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
            _socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.SendTimeout, (int)config.SendTimeout.TotalMilliseconds);
            _socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReceiveTimeout, 24 * 3600 * 1000/*one day*/);//(int)config.ReceiveTimeout.TotalMilliseconds);
            _socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.NoDelay, !config.Nagle);
            _socket.Connect(_endPoint);
            socket = _socket;
            this.inputStream = new BufferedStream(new NetworkStream(this.socket), config.BufferSize);
            Thread th = new Thread(delegate() {
                while (true)
                {
                    try
                    {
                        Receive();
                    }
                    catch (Exception ex)
                    {
                        logger.Notice(ex.Message);
                    }
                }
            });
            th.Start();
        }
        public void Reset()
        {
            try
            {
                this.inputStream.Flush();
            }
            catch { }
        }
        public bool InPool
        {
            get;
            set;
        }
        public bool Connected
        {
            get { return socket.Connected; }
        }
        public void Disconnect(bool reuseSocket)
        {
            socket.Disconnect(reuseSocket);
        }

        public int Send(byte[] data)
        {
            ThreadPool.QueueUserWorkItem(delegate(object param)
            {
                byte[] sendData = new byte[data.Length + CONST.P_START_TAG.Length + CONST.P_END_TAG.Length];
                Array.Copy(CONST.P_START_TAG, 0, sendData, 0, CONST.P_START_TAG.Length);
                Array.Copy(data, 0, sendData, CONST.P_START_TAG.Length, data.Length);
                Array.Copy(CONST.P_END_TAG, 0, sendData, CONST.P_START_TAG.Length + data.Length, CONST.P_END_TAG.Length);
                SocketError error;
                int sendcount = socket.Send(sendData, 0, sendData.Length, SocketFlags.None, out error);
                if (sendcount > 0 && error == SocketError.Success)
                {
                    return;
                }
                else if (sendcount <= 0)
                {
                    logger.Error(String.Format("Failed to write to the socket '{0}'. Error: {1}", endpoint, error));
                    throw new IOException(String.Format("Failed to write to the socket '{0}'. Error: {1}", endpoint, error));
                }
                else if (error == SocketError.TimedOut)
                {
                    logger.Warn("socket send timeout! host:" + endpoint.Address);
                    throw new SendTimeoutException();
                }
                else
                {
                    logger.Error("socket exception,error:" + error.ToString() + " Code:" + (int)error + ",host:" + endpoint.Address);
                    throw new OtherException("socket exception,error:" + error.ToString() + " Code:" + (int)error);
                }
            });
            return data.Length;
        }

        private void Receive()
        {
            Stopwatch watcher = new Stopwatch();
            watcher.Start();
            IList<byte> pData = new List<byte>();

            //find start tag
            byte[] stag = new byte[CONST.P_START_TAG.Length];
            Read(stag, 0, CONST.P_START_TAG.Length);
            if (!ArrayHelper.Equals<byte>(stag, CONST.P_START_TAG))
            {
                while (true)
                {
                    byte b = (byte)inputStream.ReadByte();
                    ArrayHelper.LeftMove<byte>(stag, 1);
                    stag[stag.Length - 1] = b;
                    if (ArrayHelper.Equals<byte>(stag, CONST.P_START_TAG))
                        break;
                }
            }

            //read version
            byte[] bVersion = new byte[SFPStruct.Version];
            Read(bVersion, 0, SFPStruct.Version);
            if (bVersion[0] == 0)
            {
                logger.Error("protocol error.");
                throw new ProtocolException();
            }
            Array.ForEach(bVersion, p => pData.Add(p));

            //read total length
            byte[] bTotalLen = new byte[SFPStruct.TotalLen];
            Read(bTotalLen, 0, SFPStruct.TotalLen);
            Array.ForEach(bTotalLen, p => pData.Add(p));
            int totalLen = BitConverter.ToInt32(bTotalLen, 0);

            //read sessionid
            byte[] bSessionId = new byte[SFPStruct.SessionId];
            Read(bSessionId, 0, SFPStruct.SessionId);
            Array.ForEach(bSessionId, p => pData.Add(p));
            int pSessionId = BitConverter.ToInt32(bSessionId, 0);

            //read other data
            byte[] otherData = new byte[totalLen - SFPStruct.Version - SFPStruct.TotalLen - SFPStruct.SessionId];
            Read(otherData, 0, otherData.Length);
            Array.ForEach(otherData, p => pData.Add(p));

            //read end tag and check
            byte[] etag = new byte[CONST.P_END_TAG.Length];
            Read(etag, 0, CONST.P_END_TAG.Length);
            if (!ArrayHelper.Equals(etag, CONST.P_END_TAG))
            {
                throw new Exception("not find end tag from stream!");
            }
            WindowData channelData;
            if (ReceiveWindows.TryGetValue(pSessionId, out channelData))
            {
                channelData.Data = pData.ToArray();
                channelData.Event.Set();
            }
            logger.Debug("total receive time:" + watcher.ElapsedMilliseconds + "ms");
        }

        public byte[] Receive(int sessionId)
        {
            AutoResetEvent ReceiveEvent = new AutoResetEvent(false);
            var channelData = new WindowData(ReceiveEvent);
            ReceiveWindows[sessionId] = channelData;
            try
            {
                if (!ReceiveEvent.WaitOne(socketConfig.ReceiveTimeout))
                {
                    throw new Com.Bj58.Spat.Gaea.Client.Exceptions.TimeoutException("Receive data timeout!");
                    
                }
            }
            finally
            {
                ReceiveWindows.Remove(sessionId);
            }
            return channelData.Data;
        }

        public void Close()
        {
            pool.Release(this);
        }

        public void Read(byte[] buffer, int offset, int count)
        {
            int read = 0;
            int shouldRead = count;

            while (read < count)
            {
                //int currentRead = this.socket.Receive(buffer, offset, shouldRead, SocketFlags.None);
                int currentRead = this.inputStream.Read(buffer, offset, shouldRead);
                if (currentRead < 1)
                    continue;
                read += currentRead;
                offset += currentRead;
                shouldRead -= currentRead;
            }
        }

        #region IDisposable 成员

        public void Dispose()
        {
            Dispose(false);
        }

        public void Dispose(bool flag)
        {
            if (flag)
            {
                pool.Destroy(this);
                inputStream.Close();
                inputStream.Dispose();
            }
            else
            {
                Close();
            }
        }

        #endregion
    }
    public class WindowData
    {
        public WindowData(AutoResetEvent even)
        {
            Event = even;
        }
        public AutoResetEvent Event
        {
            get;
            set;
        }
        public byte[] Data
        {
            get;
            set;
        }
    }
}
